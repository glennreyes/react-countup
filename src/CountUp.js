import PropTypes from 'prop-types';
import React, { useEffect, createRef, useRef } from 'react';
import warning from 'warning';
import { createCountUpInstance } from './common';

const arePropsEqual = (prevProps, nextProps) => {
  const hasCertainPropsEqual =
    prevProps.duration === nextProps.duration ||
    prevProps.end === nextProps.end ||
    prevProps.start === nextProps.start;

  return hasCertainPropsEqual || !nextProps.redraw;
};

const CountUp = React.memo(props => {
  const containerRef = createRef();
  const timer = useRef();
  const instance = useRef();

  const createInstance = () => {
    if (typeof props.children === 'function') {
      // Warn when user didn't use containerRef at all
      warning(
        containerRef.current &&
          (containerRef.current instanceof HTMLElement ||
            containerRef.current instanceof SVGTextElement),
        `Couldn't find attached element to hook the CountUp instance into! Try to attach "containerRef" from the render prop to a an HTMLElement, eg. <span ref={containerRef} />.`,
      );
    }
    return createCountUpInstance(containerRef.current, props);
  };

  useEffect(() => {
    const { children, delay } = props;
    instance.current = createInstance();

    // Don't invoke start if component is used as a render prop
    if (typeof children === 'function' && delay !== 0) return;

    // Otherwise just start immediately
    start();

    return () => {
      if (timer.current) {
        clearTimeout(timer.current);
      }
      instance.current.reset();
    };
  }, [props.duration, props.start]);

  useEffect(() => {
    // Only end value has changed, so reset and and re-animate with the updated
    // end value.

    // Don't invoke start if component is used as a render prop
    if (typeof children === 'function' && props.delay !== 0) return;

    instance.current.reset();
    instance.current.update(props.end);
  }, [props.end]);

  const pauseResume = () => {
    const { onPauseResume } = props;

    instance.current && instance.current.pauseResume();

    onPauseResume({ reset, start, update });
  };

  const reset = () => {
    const { onReset } = props;

    onReset({ pauseResume, start, update });
  };

  const restart = () => {
    reset();
    start();
  };

  const update = newEnd => {
    const { onUpdate } = props;

    instance.current && instance.current.update(newEnd);

    onUpdate({ pauseResume, reset, start });
  };

  const start = () => {
    const { delay, onEnd, onStart } = props;
    const run = () =>
      instance.current &&
      instance.current.start(() =>
        onEnd({ pauseResume, reset, start, update }),
      );

    // Delay start if delay prop is properly set
    if (delay > 0) {
      timer.current = setTimeout(run, delay * 1000);
    } else {
      run();
    }

    onStart({ pauseResume, reset, update });
  };

  const { children, className, style } = props;

  if (typeof children === 'function') {
    return children({
      countUpRef: containerRef,
      pauseResume,
      reset,
      start: restart,
      update,
    });
  }

  return <span className={className} ref={containerRef} style={style} />;
}, arePropsEqual);

CountUp.propTypes = {
  decimal: PropTypes.string,
  decimals: PropTypes.number,
  delay: PropTypes.number,
  easingFn: PropTypes.func,
  end: PropTypes.number.isRequired,
  formattingFn: PropTypes.func,
  onEnd: PropTypes.func,
  onStart: PropTypes.func,
  prefix: PropTypes.string,
  redraw: PropTypes.bool,
  separator: PropTypes.string,
  start: PropTypes.number,
  suffix: PropTypes.string,
  style: PropTypes.object,
  useEasing: PropTypes.bool,
};

CountUp.defaultProps = {
  decimal: '.',
  decimals: 0,
  delay: null,
  duration: null,
  easingFn: null,
  formattingFn: null,
  onEnd: () => {},
  onPauseResume: () => {},
  onReset: () => {},
  onStart: () => {},
  onUpdate: () => {},
  prefix: '',
  redraw: false,
  separator: '',
  start: 0,
  suffix: '',
  style: undefined,
  useEasing: true,
};

export default CountUp;
