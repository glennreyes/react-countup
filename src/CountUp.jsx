import React, { PropTypes } from 'react';
import { findDOMNode } from 'react-dom';
import shallowCompare from 'react-addons-shallow-compare';
import Count from 'countup.js';

class CountUp extends React.Component {
  componentDidMount() {
    const {
      start,
      end,
      duration,
      useEasing,
      useGrouping,
      separator,
      decimals,
      decimal,
      prefix,
      suffix,
      callback,
    } = this.props;

    this.state = {
      start,
      end,
      duration,
      useEasing,
      useGrouping,
      separator,
      decimals,
      decimal,
      prefix,
      suffix,
      callback,
    };

    this.startAnimation();
  }

  shouldComponentUpdate(nextProps, nextState) {
    return nextProps.redraw
      ? true
      : shallowCompare(this, nextProps, nextState);
  }

  componentDidUpdate() {
    this.startAnimation();
  }

  startAnimation() {
    const {
      start,
      end,
      duration,
      useEasing,
      useGrouping,
      separator,
      decimals,
      decimal,
      prefix,
      suffix,
      callback,
    } = this.props;

    const countup = new Count(findDOMNode(this), start, end, decimals, duration, {
      useEasing,
      useGrouping,
      separator,
      decimal,
      prefix,
      suffix,
    });

    countup.start(callback);
  }

  render() {
    const { className, start, style } = this.props;

    return <span className={className} style={style}>{start}</span>;
  }
}

CountUp.defaultProps = {
  start: 0,
  end: 2016,
  decimals: 0,
  duration: 4,
  useEasing: true,
  useGrouping: false,
  separator: ',',
  decimal: '.',
  prefix: '',
  suffix: '',
  redraw: false,
};

CountUp.propTypes = {
  className: PropTypes.node,
  style: PropTypes.object,
  start: PropTypes.number.isRequired,
  end: PropTypes.number.isRequired,
  decimals: PropTypes.number,
  duration: PropTypes.number,
  useEasing: PropTypes.bool,
  useGrouping: PropTypes.bool,
  separator: PropTypes.string,
  decimal: PropTypes.string,
  prefix: PropTypes.string,
  suffix: PropTypes.string,
  callback: PropTypes.func,
  redraw: PropTypes.bool,
};

export default CountUp;
