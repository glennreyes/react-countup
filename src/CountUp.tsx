import React, { Component, CSSProperties } from 'react';
import warning from 'warning';
import { createCountUpInstance } from './common';
import { CountUp as CountUpJs } from 'countup.js';
import { CallbackProps, CommonProps, RenderCounterProps } from './types';

export interface CountUpProps extends CommonProps, CallbackProps {
  className?: string;
  redraw?: boolean;
  preserveValue?: boolean;
  children?: (props: RenderCounterProps) => React.ReactNode;
  style?: CSSProperties;
}

class CountUp extends Component<CountUpProps> {
  private instance: CountUpJs | undefined;
  private timeoutId: ReturnType<typeof setTimeout> | undefined;

  static defaultProps = {
    decimal: '.',
    decimals: 0,
    delay: null,
    duration: null,
    easingFn: null,
    formattingFn: null,
    onEnd: () => {},
    onPauseResume: () => {},
    onReset: () => {},
    onStart: () => {},
    onUpdate: () => {},
    prefix: '',
    redraw: false,
    separator: '',
    start: 0,
    startOnMount: true,
    suffix: '',
    style: undefined,
    useEasing: true,
    preserveValue: false,
  };

  componentDidMount() {
    const { children, delay } = this.props;
    this.instance = this.createInstance();

    // Don't invoke start if component is used as a render prop
    if (typeof children === 'function' && delay !== 0) return;

    // Otherwise just start immediately
    this.start();
  }

  checkProps = (updatedProps: CountUpProps) => {
    const {
      start,
      suffix,
      prefix,
      redraw,
      duration,
      separator,
      decimals,
      decimal,
      className,
      formattingFn
    } = this.props;

    const hasPropsChanged =
      duration !== updatedProps.duration ||
      start !== updatedProps.start ||
      suffix !== updatedProps.suffix ||
      prefix !== updatedProps.prefix ||
      separator !== updatedProps.separator ||
      decimals !== updatedProps.decimals ||
      decimal !== updatedProps.decimal ||
      className !== updatedProps.className ||
      formattingFn !== updatedProps.formattingFn;

    return hasPropsChanged || redraw;
  };

  shouldComponentUpdate(nextProps: CountUpProps) {
    const { end } = this.props;
    return this.checkProps(nextProps) || end !== nextProps.end;
  }

  componentDidUpdate(prevProps: CountUpProps) {
    // If duration, suffix, prefix, separator or start has changed
    // there's no way to update the values.
    // So we need to re-create the CountUp instance in order to
    // restart it.
    const { end, preserveValue } = this.props;

    if (this.checkProps(prevProps)) {
      this.instance?.reset();
      this.instance = this.createInstance();
      this.start();
    }

    // Only end value has changed, so reset and and re-animate with the updated
    // end value.
    if (end !== prevProps.end) {
      if (!preserveValue) {
        this.instance?.reset();
      }
      this.instance?.update(end);
    }
  }

  componentWillUnmount() {
    if (this.timeoutId) {
      clearTimeout(this.timeoutId);
    }
    // instance.target is incorrectly marked private by typescript
    if (this.instance && (this.instance as any).target) {
      this.instance.reset();
    }
  }

  createInstance = () => {
    if (typeof this.props.children === 'function') {
      // Warn when user didn't use containerRef at all
      warning(
        this.containerRef.current instanceof Element,
        `Couldn't find attached element to hook the CountUp instance into! Try to attach "containerRef" from the render prop to a an Element, eg. <span ref={containerRef} />.`,
      );
    }
    return createCountUpInstance(this.containerRef.current, this.props);
  };

  pauseResume = () => {
    const { reset, restart: start, update } = this;
    const { onPauseResume } = this.props;

    this.instance?.pauseResume();

    onPauseResume?.({ reset, start, update });
  };

  reset = () => {
    const { pauseResume, restart: start, update } = this;
    const { onReset } = this.props;

    this.instance?.reset();

    onReset?.({ pauseResume, start, update });
  };

  restart = () => {
    this.reset();
    this.start();
  };

  start = () => {
    const { pauseResume, reset, restart: start, update } = this;
    const { delay, onEnd, onStart } = this.props;
    const run = () =>
      this.instance?.start(() => onEnd?.({ pauseResume, reset, start, update }));

    // Delay start if delay prop is properly set
    if (delay && delay > 0) {
      this.timeoutId = setTimeout(run, delay * 1000);
    } else {
      run();
    }

    onStart?.({ pauseResume, reset, update });
  };

  update = (newEnd: string | number) => {
    const { pauseResume, reset, restart: start } = this;
    const { onUpdate } = this.props;

    this.instance?.update(newEnd);

    onUpdate?.({ pauseResume, reset, start });
  };

  containerRef = React.createRef<any>();

  render() {
    const { children, className, style } = this.props;
    const { containerRef, pauseResume, reset, restart, update } = this;

    if (typeof children === 'function') {
      return children({
        countUpRef: containerRef,
        pauseResume,
        reset,
        start: restart,
        update,
      });
    }

    return <span className={className} ref={containerRef} style={style} />;
  }
}

export default CountUp;
