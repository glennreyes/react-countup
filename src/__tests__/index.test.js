import React from 'react';
import { render } from 'react-dom';
import renderer from 'react-test-renderer';
import CountUp, { formatNumber, startAnimation } from '../index';

it('renders correctly', () => {
  const createNodeMock = () => ({ startAnimation });
  const tree = renderer
    .create(<CountUp start={0} end={10} onStart={() => {}} />, {
      createNodeMock,
    })
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with wrong onStart type', () => {
  const createNodeMock = () => ({ startAnimation });
  const tree = renderer
    .create(<CountUp start={0} end={10} onStart="Something wrong" />, {
      createNodeMock,
    })
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with formattingFn', () => {
  const createNodeMock = () => ({ startAnimation });
  const tree = renderer
    .create(
      <CountUp start={0} end={10} formattingFn={v => `formated:${v}`} />,
      {
        createNodeMock,
      },
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('should update on new duration', () => {
  const node = document.createElement('div');
  const instance = render(<CountUp duration={1} />, node);
  jest.spyOn(instance, 'componentDidUpdate');
  render(<CountUp duration={2} />, node);
  expect(instance.componentDidUpdate).toHaveBeenCalled();
});

it('should update on new start', () => {
  const node = document.createElement('div');
  const instance = render(<CountUp start={1} />, node);
  jest.spyOn(instance, 'componentDidUpdate');
  render(<CountUp start={2} />, node);
  expect(instance.componentDidUpdate).toHaveBeenCalled();
});

it('should update on new end', () => {
  const node = document.createElement('div');
  const instance = render(<CountUp end={1} />, node);
  jest.spyOn(instance, 'componentDidUpdate');
  render(<CountUp end={2} />, node);
  expect(instance.componentDidUpdate).toHaveBeenCalled();
});

it('should throw an error if no component specified in startAnimation', () => {
  expect(() => startAnimation('😍')).toThrow();
});

it('formats decimal number correctly', () => {
  const formattedNumber = formatNumber(123456.789, {
    decimal: ',',
    decimals: 2,
    separator: '.',
    prefix: 'a ',
    suffix: ' b',
  });
  expect(formattedNumber).toBe('a 123.456,79 b');
});

it('formats arbitrary number correctly', () => {
  const formattedNumber = formatNumber(123456, {
    decimal: '',
    decimals: 2,
    separator: '',
    prefix: 'a ',
    suffix: ' b',
  });
  expect(formattedNumber).toBe('a 12345600 b');
});

it('formats negative number with separator correctly', () => {
  const formattedNumber = formatNumber(-123456.789, {
    decimal: '.',
    decimals: 2,
    separator: ' ',
  });
  expect(formattedNumber).toBe('-123 456.79');
});
