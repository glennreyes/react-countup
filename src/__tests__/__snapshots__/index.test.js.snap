// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<span
  className={undefined}
  style={undefined}
>
  0
</span>
`;

exports[`renders correctly with formattingFn 1`] = `
<span
  className={undefined}
  style={undefined}
>
  formated:0
</span>
`;

exports[`renders correctly with wrong onStart type 1`] = `
<span
  className={undefined}
  style={undefined}
>
  0
</span>
`;
